### PR TITLE
SONARIAC-836 Convert JSON to Minimal ARM AST Model: ARM Parse Resource

### PR DESCRIPTION
**Pull request by rudy-regazzoni-sonarsource**

### CODE REVIEW

**Comment by rudy-regazzoni-sonarsource** [comment]
Code coverage cannot be improved more: from my perspective, the only case left is a defensive condition to not crash, as I don't think we can have something else than a `ScalarTree` for a key in a JSON object.

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/ArmParser.java [hunk 1]
```diff
@@ -50,9 +67,81 @@ private static FileTree parseJson(String source, @Nullable InputFileContext inpu
...
+    extractResourcesSequence(fileTree).ifPresent(res -> {
+      List<ResourceDeclaration> resourceDeclarations = convertResources(res);
+      statements.addAll(resourceDeclarations);
+    });
+
+    return new FileImpl(statements);
+  }
+
+  private static Optional<SequenceTree> extractResourcesSequence(FileTree fileTree) {
+    MappingTree document = (MappingTree) fileTree.documents().get(0);
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/ArmParser.java:82]: Assumption: in JSON we can have only one document.

File: iac-extensions/arm/src/test/java/org/sonar/iac/arm/parser/ResourceDeclarationTest.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+      "  ]",
+      "}");
+    assertThatThrownBy(() -> parser.parse(code, null))
+      .isInstanceOf(ParseException.class)
+      .hasMessage("Expecting ScalarTree to convert to Expression, got SequenceTreeImpl");
+  }
+
+  @Test
+  void shouldThrowParseExceptionOnIncompleteResource() {
+    String[] usecases = new String[] {
```
Comment (mstachniuk) [inline iac-extensions/arm/src/test/java/org/sonar/iac/arm/parser/ResourceDeclarationTest.java:137]: It looks like a good candidate for Parameterized Tests: https://www.baeldung.com/parameterized-tests-junit-5

**Review by mstachniuk — APPROVED** [review]
LGTM and it needs some some small improvements

File: iac-extensions/arm/src/test/java/org/sonar/iac/arm/parser/ResourceDeclarationTest.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    String code = code("{",
+      "  \"resources\": [",
+      "    {",
+      "      \"type\": \"Microsoft.Kusto/clusters\",",
+      "      \"apiVersion\": \"2022-12-29\",",
+      "      \"name\": \"myResource\",",
+      "      \"location\": \"random location\",",
+      "    }",
+      "  ]",
+      "}");
```
[the remark this answers is not in the document]
Comment (rudy-regazzoni-sonarsource) [reply iac-extensions/arm/src/test/java/org/sonar/iac/arm/parser/ResourceDeclarationTest.java:87]: I went further: I merged those two tests.

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/tree/impl/ExpressionImpl.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.arm.tree.impl;
```
[the remark this answers is not in the document]
Comment (rudy-regazzoni-sonarsource) [reply iac-extensions/arm/src/main/java/org/sonar/iac/arm/tree/impl/ExpressionImpl.java:20]: I don't know how but I didn't see that the package was already there with `FileImpl` in it :sweat_smile:

**Pull request merged**